src/main/java/de/tum/in/www1/artemis/service/TextAssessmentQueueService.java

```diff
@@ -0,0 +1,167 @@
+        this.textClusterRepository = textClusterRepository;
+        this.textSubmissionService = textSubmissionService;
+    }
+
+    /**
+     * Calculates the proposedTextSubmission for a given Text exercise
+     *
+     * @param textExercise the exercise for
+     * @throws IllegalArgumentException if textExercise isn't automatically assessable
+     * @return a TextSubmission with the highest information Gain if there is one
+     */
+    public Optional<TextSubmission> getProposedTextSubmission(TextExercise textExercise) {
+        return getProposedTextSubmission(textExercise, null);
+    }
+
+    /**
+     * Calculates the proposedTextSubmission for a given Text exercise
+     *
+     * @param textExercise the exercise for
+     * @param languages  list of languages the submission which the returned  submission should have if null all languages are allowed
+     * @throws IllegalArgumentException if textExercise isn't automatically assessable
+     * @return a TextSubmission with the highest information Gain if there is one
+     */
+    @Transactional(readOnly = true)
+    public Optional<TextSubmission> getProposedTextSubmission(TextExercise textExercise, List<Language> languages) {
+
+        if (!textExercise.isAutomaticAssessmentEnabled()) {
+            throw new IllegalArgumentException("The TextExercise is not automatic assessable");
+        }
+        List<TextSubmission> textSubmissionList = textSubmissionService.getAllOpenTextSubmissions(textExercise);
+        if (textSubmissionList.isEmpty()) {
+            return Optional.empty();
+        }
+        HashMap<TextBlock, Double> smallerClusterMap = calculateSmallerClusterPercentageBatch(textSubmissionList);
+        Optional<TextSubmission> best = textSubmissionList.stream().filter(textSubmission -> languages == null || languages.contains(textSubmission.getLanguage()))
+                .max(Comparator.comparingDouble(textSubmission -> calculateInformationGain(textSubmission, smallerClusterMap)));
+        return best;
+    }
+
+    /**
+     * Given an TextSubmission calculate the expected information gain
+     * based on the smallerClusterMap and addedDistancesMap
+     *
+     * @param textSubmission the textSubmission
+     * @param smallerClusterMap Map of TextBlocks to percentage of smaller clusters the TextBlock Cluster
+     * @return information gain for the TextSubmission
+     */
+    private double calculateInformationGain(TextSubmission textSubmission, HashMap<TextBlock, Double> smallerClusterMap) {
+        List<TextBlock> textBlocks = textSubmission.getBlocks();
+        double totalScore = 0.0;
+        for (TextBlock textBlock : textBlocks) {
+            if (textBlock.isAssessable()) {
+                continue;
+            }
+            if (textBlock.getCluster() == null) {
+                continue;
+            }
+            double textBlockScore = textBlock.getAddedDistance();
+            textBlockScore /= textBlock.getCluster().size();
```

Comment:
Is it impossible that `textBlock.getCluster().size()` is 0?

Reply:
Yes it is impossible. Because textBlock is an element of textBlock.getCluster()

Reply:
```
Exception in thread "main" java.lang.ArithmeticException: / by zero
	at MyClass.main(MyClass.java:7)
```
If this is a case, it should be modularized in the objects / caught with an if statement here.